addons/website_slides_survey/controllers/slides.py

```diff
@@ -74,3 +75,34 @@ def _get_users_certificates(self, users):
             ] for user in users
         }
         return users_certificates
+
+    # Badges & Ranks Page
+    # ---------------------------------------------------
+    def _prepare_ranks_badges_values(self, **kwargs):
+        """ Extract certification badges, to render them in ranks/badges page in another section.
+        Order them by number of granted users desc and show only badges linked to opened certifications."""
+        values = super(WebsiteSlides, self)._prepare_ranks_badges_values(**kwargs)
+
+        # 1. Getting all certification badges, sorted by granted user desc
+        certification_badges = request.env['gamification.badge'].sudo().search([('survey_id', '!=', False)])
+
+        if not certification_badges:
+            return values
+
+        certification_badges = certification_badges.sorted("granted_users_count", reverse=True)
+
+        # 2. Getting all certifications slides and link certification badge and course url
+        certification_slides = request.env['slide.slide'].sudo().search([('survey_id', 'in', certification_badges.mapped('survey_id').ids)])
+        certification_badge_urls = {slide.survey_id.certification_badge_id.id: slide.channel_id.website_url for slide in certification_slides}
```

Comment:
1. You fetch all badges having a survey_id
2. You fetch all survey from these badges
3. You use the badge from these surveys

Feels like you're looping somewhere here.
Maybe it's to avoid additional requests? But it's not avoiding much since you use `.certification_badge_id` anyway.

So maybe simply do in your template:
`url = badge_id.survey_id.channel_id.website_url` ?

Or use a raw SQL request to fetch urls in an optimized way.
Or maybe I missed something here.

Reply:
On addition, it feels like you're building a dict for badges that may be filtered out below when removing state that are not 'open'. 

Maybe that filtering phase should be placed before.

Reply:
1. website_url is computed so cannot make sql query and we don't have the link from survey to slide so cannot make magic in template from badge only. :/
2. goed gezien !